From e38ddaf6e5fd6975f0e94647ae538eb2c96c3b34 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Thu, 11 Nov 2021 13:30:28 +0800
Subject: [PATCH] glamor: Fix fbo pixmap format with GL_BGRA_EXT

Red and Blue channels are swizzled in GLES path,
Fix color inverse problem with ximagesink command:
gst-launch-1.0 videotestsrc pattern=6 num-buffers=1 ! video/x-raw,format=BGRx ! videoconvert ! ximagesink display=:0 -v

This fix requires GL_EXT_texture_format_BGRA8888 extention, not applicable for upstream.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
---
 glamor/glamor.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/glamor/glamor.c b/glamor/glamor.c
index b9df95756..17d7947e1 100644
--- a/glamor/glamor.c
+++ b/glamor/glamor.c
@@ -574,9 +574,9 @@ glamor_setup_formats(ScreenPtr screen)
 
     if (glamor_priv->is_gles) {
         assert(X_BYTE_ORDER == X_LITTLE_ENDIAN);
-        glamor_add_format(screen, 24, PICT_x8b8g8r8,
+        glamor_add_format(screen, 24, PICT_x8r8g8b8,
                           GL_BGRA, GL_BGRA, GL_UNSIGNED_BYTE);
-        glamor_add_format(screen, 32, PICT_a8b8g8r8,
+        glamor_add_format(screen, 32, PICT_a8r8g8b8,
                           GL_BGRA, GL_BGRA, GL_UNSIGNED_BYTE);
     } else {
         glamor_add_format(screen, 24, PICT_x8r8g8b8,
-- 
2.17.1

